### PR TITLE
[release-4.11.0] Bug 2095226: Added changes to verify cloud connection and dhcp serviceinstance details of a  powervs instance

### PR DESCRIPTION
**Pull request by bkhadars**
Added changes for verifying the cloudconnection and dhcpserverinstance details of a powervs instance in the ibm cloud

### CODE REVIEW

File: pkg/asset/installconfig/powervs/session.go [hunk 1]
```diff
@@ -182,6 +184,37 @@ func (c *BxClient) ValidateAccountPermissions() error {
...
+//ValidateDhcpServiceInstance checks for existing dhcp instance for the provided PowerVS cloud instance
+func (c *BxClient) ValidateDhcpServiceInstance(svcInsID string) error {
+
+	//check DHCP serviceInstance
+	dhcpClient := instance.NewIBMPIDhcpClient(context.Background(), c.PISession, svcInsID)
+	dhcpServers, err := dhcpClient.GetAll()
+	if err != nil {
+		return err
+	}
+	// only one dhcp server is allowed per cloud instance
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:196]: same comment here. "only one DHCP service is allowed..."

**Review by clnperez — CHANGES_REQUESTED** [review]

File: pkg/asset/installconfig/powervs/session.go [hunk 2]
```diff
@@ -182,6 +185,67 @@ func (c *BxClient) ValidateAccountPermissions() error {
...
+		}
+		res <- len(getAllResp.CloudConnections)
+		close(res)
+		return nil
+	}()
+	for {
+		select {
+		case noOfCloudConnections := <-res:
+			if noOfCloudConnections >= 1 {
+				return fmt.Errorf("only one cloudconnection with dhcpservice is allowed per Zone")
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:240]: 1) I don't see where this checks that there is a DHCP service associated with this Cloud connection. The existence of another Cloud connection on its own isn't a problem.
2) Cloud connection is two words. All of our documentation capitalizes the first word but not the second: https://cloud.ibm.com/docs/power-iaas?topic=power-iaas-cloud-connections
3) DHCP service is two words. DHCP should always be capitalized as it is an acronym. 
4) Zone doesn't need to be capitalized. 
5) The error isn't exactly clear to an end user. Change this to something like, "only one Cloud connection can be attached to any DHCP network per account per zone."

File: pkg/asset/quota/quota.go [hunk 3]
```diff
@@ -144,7 +145,26 @@ func (a *PlatformQuotaCheck) Generate(dependencies asset.Parents) error {
...
+			return errors.Wrap(err, "failed to create bluemix client")
+		}
+
+		err = bxCli.NewPISession()
+		if err != nil {
+			return errors.Wrap(err, "failed to get PowerVS connection details")
+		}
+		err = bxCli.ValidateCloudConnectionInPowerVSRegion(context.TODO(), ic.Config.Platform.PowerVS.ServiceInstanceID)
+		if err != nil {
+			return errors.Wrap(err, "failed to meet the prerequisite for cloudConnection")
```
Comment (clnperez) [inline pkg/asset/quota/quota.go:160]: same issues with the grammar here. Cloud connections

File: pkg/asset/quota/quota.go [hunk 4]
```diff
@@ -144,7 +145,26 @@ func (a *PlatformQuotaCheck) Generate(dependencies asset.Parents) error {
...
+		}
+
+		err = bxCli.ValidateDhcpService(context.TODO(), ic.Config.Platform.PowerVS.ServiceInstanceID)
+		if err != nil {
+			return errors.Wrap(err, "failed to meet the prerequisite of dhcpservice per Power VS instance")
```
Comment (clnperez) [inline pkg/asset/quota/quota.go:165]: same here as well for DHCP

**Review by clnperez — CHANGES_REQUESTED** [review]
the code itself is fine, but the messages exposed to the user need to be cleaned up

File: pkg/asset/installconfig/powervs/session.go [hunk 5]
```diff
@@ -182,6 +185,67 @@ func (c *BxClient) ValidateAccountPermissions() error {
 	return nil
 }
 
+//ValidateDhcpService checks for existing dhcp instance for the provided PowerVS cloud instance
+func (c *BxClient) ValidateDhcpService(ctx context.Context, svcInsID string) error {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, helpers.PIGetTimeOut)
+	defer cancel()
+	res := make(chan int)
+
+	go func() error {
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/session.go:194]: With the use of context, this probably doesn't need to be an extra go routine.

File: pkg/asset/installconfig/powervs/session.go [hunk 6]
```diff
@@ -182,6 +185,67 @@ func (c *BxClient) ValidateAccountPermissions() error {
...
+		//create Power VS DHCP Client
+		dhcpClient := instance.NewIBMPIDhcpClient(ctxWithTimeout, c.PISession, svcInsID)
+		//Get all DHCP Services
+		dhcpServers, err := dhcpClient.GetAll()
+		if err != nil {
+			return err
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/session.go:200]: Wrap the errors that come from other APIs

File: pkg/asset/installconfig/powervs/session.go [hunk 7]
```diff
@@ -182,6 +185,67 @@ func (c *BxClient) ValidateAccountPermissions() error {
...
+			return nil
+		case <-ctxWithTimeout.Done():
+			return ctxWithTimeout.Err()
+		}
+	}
+}
+
+//ValidateCloudConnectionInPowerVSRegion counts cloud connection in PowerVS Region
+func (c *BxClient) ValidateCloudConnectionInPowerVSRegion(ctx context.Context, svcInsID string) error {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, helpers.PIGetTimeOut)
```
Comment (barbacbd) [inline pkg/asset/installconfig/powervs/session.go:221]: Generally the child context takes on the name ctx also. More of nit picks but it is consistent in our code. Also can look into using a specific time like 60*time.Second or whatever the equivalent is for PIGetTimeOut

File: pkg/asset/installconfig/powervs/session.go (hunk 6 above)
Comment (bkhadars) [reply pkg/asset/installconfig/powervs/session.go]: Done

File: pkg/asset/installconfig/powervs/session.go [hunk 8]
```diff
@@ -182,6 +184,59 @@ func (c *BxClient) ValidateAccountPermissions() error {
...
+	getAllResp, err := cloudConnectionClient.GetAll()
+	if err != nil {
+		return errors.Wrap(err, "failed to get existing Cloud connection details")
+	}
+	for _, cc := range getAllResp.CloudConnections {
+		cloudConnectionsIDs = append(cloudConnectionsIDs, *cc.CloudConnectionID)
+	}
+
+	//check for DHCP Service attached to Cloud connection
+	if len(cloudConnectionsIDs) >= 1 {
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:222]: can you omit this check? i could be wrong but i think that the range loop just won't do anything if there's nothing there, so you don't need this

Comment (bkhadars) [reply pkg/asset/installconfig/powervs/session.go]: @clnperez As discussed I have removed the above check and added the check to verify number of cloud connection count per zone..

File: pkg/asset/installconfig/powervs/session.go [hunk 9]
```diff
@@ -182,6 +184,62 @@ func (c *BxClient) ValidateAccountPermissions() error {
...
+	cloudConnectionClient := instance.NewIBMPICloudConnectionClient(ctx, c.PISession, svcInsID)
+
+	//check number of cloudconnections
+	getAllResp, err := cloudConnectionClient.GetAll()
+	if err != nil {
+		return errors.Wrap(err, "failed to get existing Cloud connection details")
+	}
+
+	if len(getAllResp.CloudConnections) >= 2 {
+		return fmt.Errorf("cannot create new cloud connection in powervs zone,only 2 cloud connections allowed")
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:219]: i'm sorry to keep picking on grammar, but this is a run-on sentence and needs proper nouns to be capitalized. "cannot create new Cloud connection in Power VS. Only two Cloud connections are allowed per zone."

**Pull request merged**